File: LeetCode/remote/house_robber.rb
# You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security system connected and it will automatically contact the police if two adjacent houses were broken into on the same night.
#
# Given a list of non-negative integers representing the amount of money of each house, determine the maximum amount of money you can rob tonight without alerting the police.
#

require 'test/unit'
extend Test::Unit::Assertions

def rob(nums)
    last = now = 0
    pre = {}
    nums.each { |x|
        if last + x >= now 
            max = last + x
        else
            max = now
        end
        last = now
        now = max
    }

    puts result.inspect
    return now
end

assert_equal(25, rob([1, 2, 3, 4, 5, 6, 7, 8, 9]))

# No idea what was going on above..
# https://leetcode.com/submissions/detail/57718789/
#
# Submission Details
# 69 / 69 test cases passed.
#   Status: Accepted
#   Runtime: 108 ms
#       
#       Submitted: 1 minute ago
#
def rob(nums, hash = nil)
    hash ||= {}
    return 0 if nums.nil? || nums.empty?
    return hash[nums] if hash[nums]
    hash[nums] = [nums.first + rob(nums.drop(2), hash), rob(nums.drop(1), hash)].max
end

File: LeetCode/remote/house_robber.cs
// https://leetcode.com/problems/house-robber/
//
// You are a professional robber planning to rob houses along a street.
// Each house has a certain amount of money stashed,
// the only constraint stopping you from robbing each of them is that
// adjacent houses have security system connected and it will automatically
// contact the police if two adjacent houses were broken into on the same night.
//
// Given a list of non-negative integers representing the amount of money of each house,
// determine the maximum amount of money you can rob tonight without alerting the police.
//
//
// Submission Details
// 69 / 69 test cases passed.
//  Status: Accepted
//  Runtime: 164 ms
//      
//      Submitted: 0 minutes ago
//

using System;

public class Solution {

    // 
    //     M(k) = money at the kth house
    //     P(0) = 0
    //     P(1) = M(1)
    //     P(k) = max(P(k−2) + M(k), P(k−1))
    //
    public int Rob(int[] nums) {
        if (nums.Length == 0)
        {
            return 0;
        }

        if (nums.Length == 1)
        {
            return nums[0];
        }

        var s1 = 0;
        var s2 = 0;
        for (var i = 0; i < nums.Length; i++)
        {
            s1 = Math.Max(s1 + nums[i], s2);
            var c = s1; s1 = s2; s2 = c;
        }

        return s2;
    }

    static void Main()
    {
        var c = new Solution();
        Console.WriteLine(c.Rob(new [] { 1, 2, 3, 4, 5, 6 }));
    }

    //  https://leetcode.com/submissions/detail/62518565/
    //
    //  Submission Details
    //  69 / 69 test cases passed.
    //      Status: Accepted
    //      Runtime: 168 ms
    //          
    //          Submitted: 0 minutes ago
    public int Rob(int[] nums) {
        return nums.Aggregate(
                Tuple.Create(0, 0),
                (acc, x) => {
                    return Tuple.Create(acc.Item2, Math.Max(x + acc.Item1, acc.Item2));
                }).Item2;
    }
}
